File: Data Structure/Linked List/Medium/LET_no430_Flatten a Multilevel Doubly Linked List.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* prev;
    Node* next;
    Node* child;
};
*/

class Solution {
public:
    Node* flatten(Node* head) {
        for (Node* cur = head; cur; cur = cur->next) {
            if (cur->child) {
                Node* nxt = cur->next;
                cur->next = cur->child;
                cur->child->prev = cur;
                cur->child = nullptr;
                Node* nxt2 = cur->next;
                while (nxt2->next)
                    nxt2 = nxt2->next;
                nxt2->next = nxt;
                if (nxt)
                    nxt->prev = nxt2;
            }
        }
        
        return head;
    }
};

File: Data Structure/Linked List/Medium/LET_no430_Flatten a Multilevel Doubly Linked List.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val, prev, next, child):
        self.val = val
        self.prev = prev
        self.next = next
        self.child = child
"""

class Solution:
    def flatten(self, head: 'Node') -> 'Node':
        cur = head
        while cur:
            if cur.child:
                nxt = cur.next
                cur.next = cur.child
                cur.child = None
                cur.next.prev = cur
                nxt2 = cur.next
                while nxt2.next:
                    nxt2 = nxt2.next
                nxt2.next = nxt
                if nxt:
                    nxt.prev = nxt2
            cur = cur.next
        return head